collect_app/src/main/java/org/odk/collect/android/widgets/BaseStringWidget.java

```diff
@@ -81,7 +81,15 @@ protected void registerToClearAnswerOnLongPress(FormEntryActivity activity) {
         }
     }
 
-    public abstract void setDisplayValueFromModel();
+    public void setDisplayValueFromModel() {
```

Comment:
Yes, polymorphism is what you want. `instanceof` generally is a last-resort kind of solution. Here you're in charge of the design so you can easily avoid it by adding another level to your hierarchy.